src/Avalonia.Controls/Primitives/PopupPositioning/IPopupPositioner.cs

```diff
@@ -464,49 +464,46 @@ static class PopupPositionerExtensions
-                    positionerParameters.Gravity = PopupGravity.BottomRight;
-                }
-                else if (placement == PlacementMode.Bottom)
-                {
-                    positionerParameters.Anchor = PopupAnchor.BottomLeft;
-                    positionerParameters.Gravity = PopupGravity.BottomRight;
-                }
-                else if (placement == PlacementMode.Left)
-                {
-                    positionerParameters.Anchor = PopupAnchor.TopLeft;
-                    positionerParameters.Gravity = PopupGravity.BottomLeft;
-                }
-                else if (placement == PlacementMode.Top)
+                var targetSize = positionerParameters.Size;
+                if (targetSize != default)
                 {
-                    positionerParameters.Anchor = PopupAnchor.TopLeft;
-                    positionerParameters.Gravity = PopupGravity.TopRight;
+                    positionerParameters.Offset += new Point(
+                        (anchorRectangle.Width - targetSize.Width) / 2,
+                        (anchorRectangle.Height - targetSize.Height) / 2);
                 }
-                else if (placement == PlacementMode.AnchorAndGravity)
+            }
+            else
+            {
+                positionerParameters.AnchorRectangle = GetAnchorRectangle(topLevel, target, rect);
+
+                var parameters = placement switch
                 {
-                    positionerParameters.Anchor = anchor;
-                    positionerParameters.Gravity = gravity;
-                }
-                else
-                    throw new InvalidOperationException("Invalid value for Popup.PlacementMode");
+                    PlacementMode.Bottom => (PopupAnchor.Bottom, PopupGravity.Bottom),
+                    PlacementMode.Right => (PopupAnchor.Right, PopupGravity.Right),
+                    PlacementMode.Left => (PopupAnchor.Left, PopupGravity.Left),
+                    PlacementMode.Top => (PopupAnchor.Top, PopupGravity.Top),
+                    PlacementMode.AnchorAndGravity => (anchor, gravity),
+                    PlacementMode.TopEdgeAlignedRight => (PopupAnchor.TopRight, PopupGravity.TopLeft),
+                    PlacementMode.TopEdgeAlignedLeft => (PopupAnchor.TopLeft, PopupGravity.TopRight),
+                    PlacementMode.BottomEdgeAlignedLeft => (PopupAnchor.BottomLeft, PopupGravity.BottomRight),
+                    PlacementMode.BottomEdgeAlignedRight => (PopupAnchor.BottomRight, PopupGravity.BottomLeft),
+                    PlacementMode.LeftEdgeAlignedTop => (PopupAnchor.TopLeft, PopupGravity.BottomLeft),
+                    PlacementMode.LeftEdgeAlignedBottom => (PopupAnchor.BottomLeft, PopupGravity.TopLeft),
+                    PlacementMode.RightEdgeAlignedTop => (PopupAnchor.TopRight, PopupGravity.BottomRight),
+                    PlacementMode.RightEdgeAlignedBottom => (PopupAnchor.BottomRight, PopupGravity.TopRight),
```

Comment:
@amwx specific placement calculation algorithm is different here, as I don't set PlacementRect for any of these. But visually it has the same effect as before.

Reply:
As long as the behavior is the same its fine. I don't remember why I had the Rect stuff there too (it might've been old popup code didn't work right and has been fixed since?)